node/primitives/src/lib.rs

```diff
@@ -227,12 +227,23 @@ pub type UncheckedSignedFullStatement = UncheckedSigned<Statement, CompactStatem
 /// Candidate invalidity details
 #[derive(Debug)]
 pub enum InvalidCandidate {
-	/// Failed to execute.`validate_block`. This includes function panicking.
+	/// Failed to prevalidate.
+	PrevalidationError(String),
+	/// Failed to compile.
+	CompilationError(String),
+	/// Failed to prepare in some other way. This includes function panicking.
+	PreparationError(String),
+	/// Failed to execute `validate_block`. This includes function panicking.
 	ExecutionError(String),
 	/// Validation outputs check doesn't pass.
 	InvalidOutputs,
+	// TODO: Currently unused. Preparation timeouts are treated as non-deterministic, so this can
+	// never be instantiated, whereas execution timeouts are deterministic. Should this
+	// inconsistency be addressed?
```

Comment:
I noticed that preparation timeouts are treated as non-deterministic (potentially spurious, reported as `ValidationFailed`) whereas execution timeouts are considered deterministic (reported as `InvalidCandidate`). This might make sense if pre-checking is supposed to rule out bad PVFs, haven't looked into it TBH.